From patchwork Wed Apr  7 12:09:22 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Max Gurtovoy <mgurtovoy@nvidia.com>
X-Patchwork-Id: 12187925
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.8 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 8FB59C433ED
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:37 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 57001610FB
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:37 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S235259AbhDGMJn (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Wed, 7 Apr 2021 08:09:43 -0400
Received: from mail-dm6nam12on2042.outbound.protection.outlook.com
 ([40.107.243.42]:13795
        "EHLO NAM12-DM6-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S234810AbhDGMJj (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 7 Apr 2021 08:09:39 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=W+iSFKK/47tly0COdPOrxyRWNi1d9AbaYm47IUSVeg8hSqgH52bj+lLJXw/oAo6mEiTDK3TgXr9jkjmSo2ClDkpV771XTsukZDPadLB69J20Y6DueEmN4kKnGVTh9Y+oVcQiDYZgao8XfwEyUadUY9E2R77tIE1NdLMeCu4hLDxOmu7Nd+sCWMN+yypE/ma6SJjI/aa9Zwelf27Yxs5khdk0S1iuU/tckW+0jaB5HyFNbfG9LhgksB/pGlFwOn4Ips2f73sYVXoXyHI7wBfbmRSo/1j7tDFkd0F8ZjKPXYYFvIVlgv7b7rGoDjD9ATobY3lP5YR2pRjE28MLK3s4Og==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=yg4L4+RG827wx7kvUQi3gfqCTJEpv9Ob8nuvTmw4NVw=;
 b=VzFmWQLUopzTq0GjnrBsiCWc/qPC5ENl5XQY3wb/NzCGXXw+H/llL05c88pcC3V7CB8+cAjx753FoKhyLFzNuNghuUy8BZ9BnjvHGyRWJ+UhxWRhMqU/K1WdszqHTBltd3ATe18jGwUVdVX5by/Qm3wKezPQFcFoK1CesWIhlhbN9siOivCEPLsohV82QUOcVUUckUDJ932+HFpHKChIr3OVFLJHA9N69PWCxSchoZ2dgMJvTzGLCVXVR648hlEyzIasHx9xtl5SNICcvJ01J6qtaVqiu4+OqYnnOddr+4oSaptxuOpOgxHbhKQ59QgWfADGR0vOW4M8/aTgeNLsqg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass (sender ip is
 216.228.112.35) smtp.rcpttodomain=lists.linux-foundation.org
 smtp.mailfrom=nvidia.com; dmarc=pass (p=none sp=none pct=100) action=none
 header.from=nvidia.com; dkim=none (message not signed); arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=Nvidia.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=yg4L4+RG827wx7kvUQi3gfqCTJEpv9Ob8nuvTmw4NVw=;
 b=dRbxa/raRtFPLgjIbwezOwso1kYOqNR+LXdxSX4meoRHo//A9tPkGUHR/mjccYL7Qr8ppbFIZ3n0GeZN1ptNxNULViswVz8JHiiOfBBoBnZg29viXIL5+OEKFO3BQvEMkVJJMOsLV3wGNsmPTcsS7HhaMMZMNdN4fCoaLmr2MRrOLE5/Ioi8KYx4222dfslOirUD559kdEK+LF9+kgLl+5PUSk8teXv+zDHKP4gmcpkETeKS56BAB7Ld5PcoTsc/50rgR3cmn/ekns8C+OVUJAtTzsQ7nd9Ror3J+w5KvDPWGu3a7JH2sZSqWae9nxEqaO37YZyWrPC/vqHy0l11oA==
Received: from DS7PR05CA0025.namprd05.prod.outlook.com (2603:10b6:5:3b9::30)
 by DM5PR12MB1530.namprd12.prod.outlook.com (2603:10b6:4:4::23) with Microsoft
 SMTP Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.3999.27; Wed, 7 Apr 2021 12:09:28 +0000
Received: from DM6NAM11FT037.eop-nam11.prod.protection.outlook.com
 (2603:10b6:5:3b9:cafe::34) by DS7PR05CA0025.outlook.office365.com
 (2603:10b6:5:3b9::30) with Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4020.9 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:28 +0000
X-MS-Exchange-Authentication-Results: spf=pass (sender IP is 216.228.112.35)
 smtp.mailfrom=nvidia.com; lists.linux-foundation.org; dkim=none (message not
 signed) header.d=none;lists.linux-foundation.org; dmarc=pass action=none
 header.from=nvidia.com;
Received-SPF: Pass (protection.outlook.com: domain of nvidia.com designates
 216.228.112.35 as permitted sender) receiver=protection.outlook.com;
 client-ip=216.228.112.35; helo=mail.nvidia.com;
Received: from mail.nvidia.com (216.228.112.35) by
 DM6NAM11FT037.mail.protection.outlook.com (10.13.172.122) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384) id
 15.20.4020.17 via Frontend Transport; Wed, 7 Apr 2021 12:09:27 +0000
Received: from HQMAIL101.nvidia.com (172.20.187.10) by HQMAIL111.nvidia.com
 (172.20.187.18) with Microsoft SMTP Server (TLS) id 15.0.1497.2; Wed, 7 Apr
 2021 12:09:27 +0000
Received: from r-nvmx02.mtr.labs.mlnx (172.20.145.6) by mail.nvidia.com
 (172.20.187.10) with Microsoft SMTP Server id 15.0.1497.2 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:25 +0000
From: Max Gurtovoy <mgurtovoy@nvidia.com>
To: <mst@redhat.com>, <kvm@vger.kernel.org>,
        <virtualization@lists.linux-foundation.org>, <jasowang@redhat.com>
CC: <oren@nvidia.com>, <nitzanc@nvidia.com>,
        Max Gurtovoy <mgurtovoy@nvidia.com>
Subject: [PATCH 1/3] virtio: update reset callback to return status
Date: Wed, 7 Apr 2021 12:09:22 +0000
Message-ID: <20210407120924.133294-1-mgurtovoy@nvidia.com>
X-Mailer: git-send-email 2.25.4
MIME-Version: 1.0
X-EOPAttributedMessage: 0
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 8705d255-7db6-4be8-1a6d-08d8f9bdfde6
X-MS-TrafficTypeDiagnostic: DM5PR12MB1530:
X-Microsoft-Antispam-PRVS: 
 <DM5PR12MB1530541F95B1F4A8AAC8759DDE759@DM5PR12MB1530.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:2043;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 YByB3DotBcRpeaMBOFVBkCvXL2G70bZlCpv9Mnuo4t6nB/TSRY+DNHc6XSFoMsyFS7DgDIl52kRUsgrE21HWsxrR+dQOvTbKMxgB0SsuXuSti8Oz0eChxPDzFcMEqcQlwG8tuUqjwu4C2htksD+Eys9JW8N/g1xu+tVObXAssCw/YLWAkDYtMWR+vbY6Ap9XlARiuLXVrc6e+HJpluL1NWoKrjNbXfoj7tyZ4n1CmUSKmw7Pjt+ddRQoSOITTUigwtGZitVSMRdZa9ulx9olfn0TemE5zOjFhQ+HY7autShb9c8S6UECt8HsM3TPMiDOmt51egqmJA0G0M84kgtOCfMhH+xVGQQpITqgqQtROsvpiWW3JzcDWxnt+mWam1pSpIWTlFfIC++N6YJyTTUcwnS7dO850JnauUiK02elzxVtIiMJvx7VOM3S/F+DE0pMEv7P5mMB/+Qpj35JmDPpeiGgvM2BFDyaLBHbkh8JnaU4Rang9no7eOEifj/lObE2fR8g7kc66i8fjc20H3TyWNomjWhliCJj7q3yeu1/JAafbpwVIzk1dKEGRf/dnqut9PzoIuNsWZbj/yPlIFpg8r7K6EWdthQzPD45QDI4+mhMJjs3wi2GdvbSUzXXIrHTs4x5l2soY9bsb9kvYrVV9x5ByaP2XOlxdcY5rrypCMs=
X-Forefront-Antispam-Report: 
 CIP:216.228.112.35;CTRY:US;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:mail.nvidia.com;PTR:schybrid02.nvidia.com;CAT:NONE;SFS:(4636009)(136003)(376002)(346002)(396003)(39860400002)(36840700001)(46966006)(316002)(6666004)(26005)(8936002)(2906002)(8676002)(36906005)(5660300002)(82740400003)(54906003)(110136005)(356005)(47076005)(36860700001)(86362001)(70206006)(83380400001)(70586007)(1076003)(478600001)(82310400003)(107886003)(7636003)(4326008)(36756003)(2616005)(336012)(186003)(426003);DIR:OUT;SFP:1101;
X-OriginatorOrg: Nvidia.com
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 07 Apr 2021 12:09:27.5596
 (UTC)
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 8705d255-7db6-4be8-1a6d-08d8f9bdfde6
X-MS-Exchange-CrossTenant-Id: 43083d15-7273-40c1-b7db-39efd9ccc17a
X-MS-Exchange-CrossTenant-OriginalAttributedTenantConnectingIp: 
 TenantId=43083d15-7273-40c1-b7db-39efd9ccc17a;Ip=[216.228.112.35];Helo=[mail.nvidia.com]
X-MS-Exchange-CrossTenant-AuthSource: 
 DM6NAM11FT037.eop-nam11.prod.protection.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Anonymous
X-MS-Exchange-CrossTenant-FromEntityHeader: HybridOnPrem
X-MS-Exchange-Transport-CrossTenantHeadersStamped: DM5PR12MB1530
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

The reset device operation, usually is an operation that might fail from
various reasons. For example, the controller might be in a bad state and
can't answer to any request. Usually, the paravirt SW based virtio
devices always succeed in reset operation but this is not the case for
HW based virtio devices.

This commit is also a preparation for adding a timeout mechanism for
resetting virtio devices.

Signed-off-by: Max Gurtovoy <mgurtovoy@nvidia.com>
Reported-by: kernel test robot <lkp@intel.com>
---
 drivers/remoteproc/remoteproc_virtio.c |  3 ++-
 drivers/virtio/virtio.c                | 22 +++++++++++++++-------
 drivers/virtio/virtio_mmio.c           |  3 ++-
 drivers/virtio/virtio_pci_legacy.c     |  3 ++-
 drivers/virtio/virtio_pci_modern.c     |  3 ++-
 drivers/virtio/virtio_vdpa.c           |  3 ++-
 include/linux/virtio_config.h          |  5 +++--
 7 files changed, 28 insertions(+), 14 deletions(-)

diff --git a/drivers/remoteproc/remoteproc_virtio.c b/drivers/remoteproc/remoteproc_virtio.c
index 0cc617f76068..432f39288c15 100644
--- a/drivers/remoteproc/remoteproc_virtio.c
+++ b/drivers/remoteproc/remoteproc_virtio.c
@@ -191,7 +191,7 @@ static void rproc_virtio_set_status(struct virtio_device *vdev, u8 status)
 	dev_dbg(&vdev->dev, "status: %d\n", status);
 }
 
-static void rproc_virtio_reset(struct virtio_device *vdev)
+static int rproc_virtio_reset(struct virtio_device *vdev)
 {
 	struct rproc_vdev *rvdev = vdev_to_rvdev(vdev);
 	struct fw_rsc_vdev *rsc;
@@ -200,6 +200,7 @@ static void rproc_virtio_reset(struct virtio_device *vdev)
 
 	rsc->status = 0;
 	dev_dbg(&vdev->dev, "reset !\n");
+	return 0;
 }
 
 /* provide the vdev features as retrieved from the firmware */
diff --git a/drivers/virtio/virtio.c b/drivers/virtio/virtio.c
index 4b15c00c0a0a..ddbfd5b5f3bd 100644
--- a/drivers/virtio/virtio.c
+++ b/drivers/virtio/virtio.c
@@ -338,7 +338,7 @@ int register_virtio_device(struct virtio_device *dev)
 	/* Assign a unique device index and hence name. */
 	err = ida_simple_get(&virtio_index_ida, 0, 0, GFP_KERNEL);
 	if (err < 0)
-		goto out;
+		goto out_err;
 
 	dev->index = err;
 	dev_set_name(&dev->dev, "virtio%u", dev->index);
@@ -349,7 +349,9 @@ int register_virtio_device(struct virtio_device *dev)
 
 	/* We always start by resetting the device, in case a previous
 	 * driver messed it up.  This also tests that code path a little. */
-	dev->config->reset(dev);
+	err = dev->config->reset(dev);
+	if (err)
+		goto out_ida;
 
 	/* Acknowledge that we've seen the device. */
 	virtio_add_status(dev, VIRTIO_CONFIG_S_ACKNOWLEDGE);
@@ -362,10 +364,14 @@ int register_virtio_device(struct virtio_device *dev)
 	 */
 	err = device_add(&dev->dev);
 	if (err)
-		ida_simple_remove(&virtio_index_ida, dev->index);
-out:
-	if (err)
-		virtio_add_status(dev, VIRTIO_CONFIG_S_FAILED);
+		goto out_ida;
+
+	return 0;
+
+out_ida:
+	ida_simple_remove(&virtio_index_ida, dev->index);
+out_err:
+	virtio_add_status(dev, VIRTIO_CONFIG_S_FAILED);
 	return err;
 }
 EXPORT_SYMBOL_GPL(register_virtio_device);
@@ -408,7 +414,9 @@ int virtio_device_restore(struct virtio_device *dev)
 
 	/* We always start by resetting the device, in case a previous
 	 * driver messed it up. */
-	dev->config->reset(dev);
+	ret = dev->config->reset(dev);
+	if (ret)
+		goto err;
 
 	/* Acknowledge that we've seen the device. */
 	virtio_add_status(dev, VIRTIO_CONFIG_S_ACKNOWLEDGE);
diff --git a/drivers/virtio/virtio_mmio.c b/drivers/virtio/virtio_mmio.c
index 56128b9c46eb..12b8f048c48d 100644
--- a/drivers/virtio/virtio_mmio.c
+++ b/drivers/virtio/virtio_mmio.c
@@ -256,12 +256,13 @@ static void vm_set_status(struct virtio_device *vdev, u8 status)
 	writel(status, vm_dev->base + VIRTIO_MMIO_STATUS);
 }
 
-static void vm_reset(struct virtio_device *vdev)
+static int vm_reset(struct virtio_device *vdev)
 {
 	struct virtio_mmio_device *vm_dev = to_virtio_mmio_device(vdev);
 
 	/* 0 status means a reset. */
 	writel(0, vm_dev->base + VIRTIO_MMIO_STATUS);
+	return 0;
 }
 
 
diff --git a/drivers/virtio/virtio_pci_legacy.c b/drivers/virtio/virtio_pci_legacy.c
index d62e9835aeec..ff4d9506aa65 100644
--- a/drivers/virtio/virtio_pci_legacy.c
+++ b/drivers/virtio/virtio_pci_legacy.c
@@ -89,7 +89,7 @@ static void vp_set_status(struct virtio_device *vdev, u8 status)
 	iowrite8(status, vp_dev->ioaddr + VIRTIO_PCI_STATUS);
 }
 
-static void vp_reset(struct virtio_device *vdev)
+static int vp_reset(struct virtio_device *vdev)
 {
 	struct virtio_pci_device *vp_dev = to_vp_device(vdev);
 	/* 0 status means a reset. */
@@ -99,6 +99,7 @@ static void vp_reset(struct virtio_device *vdev)
 	ioread8(vp_dev->ioaddr + VIRTIO_PCI_STATUS);
 	/* Flush pending VQ/configuration callbacks. */
 	vp_synchronize_vectors(vdev);
+	return 0;
 }
 
 static u16 vp_config_vector(struct virtio_pci_device *vp_dev, u16 vector)
diff --git a/drivers/virtio/virtio_pci_modern.c b/drivers/virtio/virtio_pci_modern.c
index fbd4ebc00eb6..cc3412a96a17 100644
--- a/drivers/virtio/virtio_pci_modern.c
+++ b/drivers/virtio/virtio_pci_modern.c
@@ -158,7 +158,7 @@ static void vp_set_status(struct virtio_device *vdev, u8 status)
 	vp_modern_set_status(&vp_dev->mdev, status);
 }
 
-static void vp_reset(struct virtio_device *vdev)
+static int vp_reset(struct virtio_device *vdev)
 {
 	struct virtio_pci_device *vp_dev = to_vp_device(vdev);
 	struct virtio_pci_modern_device *mdev = &vp_dev->mdev;
@@ -174,6 +174,7 @@ static void vp_reset(struct virtio_device *vdev)
 		msleep(1);
 	/* Flush pending VQ/configuration callbacks. */
 	vp_synchronize_vectors(vdev);
+	return 0;
 }
 
 static u16 vp_config_vector(struct virtio_pci_device *vp_dev, u16 vector)
diff --git a/drivers/virtio/virtio_vdpa.c b/drivers/virtio/virtio_vdpa.c
index e28acf482e0c..da76190f7b07 100644
--- a/drivers/virtio/virtio_vdpa.c
+++ b/drivers/virtio/virtio_vdpa.c
@@ -97,11 +97,12 @@ static void virtio_vdpa_set_status(struct virtio_device *vdev, u8 status)
 	return ops->set_status(vdpa, status);
 }
 
-static void virtio_vdpa_reset(struct virtio_device *vdev)
+static int virtio_vdpa_reset(struct virtio_device *vdev)
 {
 	struct vdpa_device *vdpa = vd_get_vdpa(vdev);
 
 	vdpa_reset(vdpa);
+	return 0;
 }
 
 static bool virtio_vdpa_notify(struct virtqueue *vq)
diff --git a/include/linux/virtio_config.h b/include/linux/virtio_config.h
index 8519b3ae5d52..d2b0f1699a75 100644
--- a/include/linux/virtio_config.h
+++ b/include/linux/virtio_config.h
@@ -44,9 +44,10 @@ struct virtio_shm_region {
  *	status: the new status byte
  * @reset: reset the device
  *	vdev: the virtio device
- *	After this, status and feature negotiation must be done again
+ *	Upon success, status and feature negotiation must be done again
  *	Device must not be reset from its vq/config callbacks, or in
  *	parallel with being added/removed.
+ *	Returns 0 on success or error status.
  * @find_vqs: find virtqueues and instantiate them.
  *	vdev: the virtio_device
  *	nvqs: the number of virtqueues to find
@@ -82,7 +83,7 @@ struct virtio_config_ops {
 	u32 (*generation)(struct virtio_device *vdev);
 	u8 (*get_status)(struct virtio_device *vdev);
 	void (*set_status)(struct virtio_device *vdev, u8 status);
-	void (*reset)(struct virtio_device *vdev);
+	int (*reset)(struct virtio_device *vdev);
 	int (*find_vqs)(struct virtio_device *, unsigned nvqs,
 			struct virtqueue *vqs[], vq_callback_t *callbacks[],
 			const char * const names[], const bool *ctx,

From patchwork Wed Apr  7 12:09:23 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Max Gurtovoy <mgurtovoy@nvidia.com>
X-Patchwork-Id: 12187927
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.8 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id C04E8C433ED
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:42 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 968B0610FB
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:42 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S236844AbhDGMJs (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Wed, 7 Apr 2021 08:09:48 -0400
Received: from mail-dm3nam07on2060.outbound.protection.outlook.com
 ([40.107.95.60]:17536
        "EHLO NAM02-DM3-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S235093AbhDGMJl (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 7 Apr 2021 08:09:41 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=UIYaZhq/6W20fnA0qgYD15oAdmA9t4TInEEqG9+V1ACyUwTfl0K5pCZY2cQWaNj/lLQ5UFCXHz+AoSDnztjfAn5cf4C/qV3WPi1rPhykkuWO6ylvUbQd2Vh4gT3jSOKkLTgzyd40H8wqY2NPYU2QJVzdU+1ySacH1QYX3c5tzWlFIRhdW0Qii5bCR6X1HMMPDgb8WqJf5yLLhwCq07WXIGP8qNqZSldh/p6F00fAnOFD/aZVZvPcO0/Po9FYVz+ed9LT93ASCB/hb0J6LYSpK8pOBtNAsSB0SiWPrANIUr+fypMmapMv5NhZQoDpxLzzrmcXAtqKHvLVpXoht040PA==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=a0EBD+tNxua7zQadLmEJYcvQW9eBq9/zvN7Jf2FQjKQ=;
 b=IrAYkSeAoVwjkIQA3M2OTi3ny/yudbZCUFeYxvUtbZSHvYi9jdaLjT29HxxNWnH+tfjCtpd7cDYAEJYXWliXkDcJVWvLInW3iUx7b//94+FTbzpEoAHpgWdccnC9FID9GjeviNMDLKE0wMLNBuWF7zLJZ9V6C9ranCXArxmQ6ge5IXan++9RyJTq8ojuGcT/nFkFNpV44fPnE8EC7AhBf+QJfoGSn7LRcaLXMYdMMV9R+qucKc4NClukdB7Nk7C3yHDxHNGe6fNlEU9dZK0lLTGTwKsXIQeLgzFzfdEadqF1LuW4oADjJ+kOxKnIO70HUHcygeery/rvDoAwj52o9w==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass (sender ip is
 216.228.112.35) smtp.rcpttodomain=lists.linux-foundation.org
 smtp.mailfrom=nvidia.com; dmarc=pass (p=none sp=none pct=100) action=none
 header.from=nvidia.com; dkim=none (message not signed); arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=Nvidia.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=a0EBD+tNxua7zQadLmEJYcvQW9eBq9/zvN7Jf2FQjKQ=;
 b=pStg1+X4mArKwrbBfC67ocfHWzBRMKlcu12+Ty1WYcPBLgI7x57+EtmzXHw+yQsVcVWb02Sai/+4YIKrE8Fz1ltq1+omaekXiVTa0/tEXiDddUTzk3h0OCS7ag+lfLMjloo6n+sFgv5k6nJ8SBOHzZyVmfwB4knujNXOfGDjiJByvhTbhVTKW8C5IQPIFHUIl6qmB0A7V6yuEsniGnSibKD+WHzmVcjCet2hj9kh41CdQNGBUFoTfzD7hnJDkkmGbkOWDS8xABrAUnsXvssIB13N9pUWXpig5SYooGriw1oELEJPVcZ6ZLA7KZG0S2ZbpD6W+lQnQubyTRQVN33NaA==
Received: from DM5PR06CA0081.namprd06.prod.outlook.com (2603:10b6:3:4::19) by
 BYAPR12MB4696.namprd12.prod.outlook.com (2603:10b6:a03:9d::15) with Microsoft
 SMTP Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.3999.28; Wed, 7 Apr 2021 12:09:30 +0000
Received: from DM6NAM11FT004.eop-nam11.prod.protection.outlook.com
 (2603:10b6:3:4:cafe::78) by DM5PR06CA0081.outlook.office365.com
 (2603:10b6:3:4::19) with Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4020.17 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:30 +0000
X-MS-Exchange-Authentication-Results: spf=pass (sender IP is 216.228.112.35)
 smtp.mailfrom=nvidia.com; lists.linux-foundation.org; dkim=none (message not
 signed) header.d=none;lists.linux-foundation.org; dmarc=pass action=none
 header.from=nvidia.com;
Received-SPF: Pass (protection.outlook.com: domain of nvidia.com designates
 216.228.112.35 as permitted sender) receiver=protection.outlook.com;
 client-ip=216.228.112.35; helo=mail.nvidia.com;
Received: from mail.nvidia.com (216.228.112.35) by
 DM6NAM11FT004.mail.protection.outlook.com (10.13.172.217) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384) id
 15.20.4020.17 via Frontend Transport; Wed, 7 Apr 2021 12:09:29 +0000
Received: from HQMAIL101.nvidia.com (172.20.187.10) by HQMAIL111.nvidia.com
 (172.20.187.18) with Microsoft SMTP Server (TLS) id 15.0.1497.2; Wed, 7 Apr
 2021 12:09:29 +0000
Received: from r-nvmx02.mtr.labs.mlnx (172.20.145.6) by mail.nvidia.com
 (172.20.187.10) with Microsoft SMTP Server id 15.0.1497.2 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:27 +0000
From: Max Gurtovoy <mgurtovoy@nvidia.com>
To: <mst@redhat.com>, <kvm@vger.kernel.org>,
        <virtualization@lists.linux-foundation.org>, <jasowang@redhat.com>
CC: <oren@nvidia.com>, <nitzanc@nvidia.com>,
        Max Gurtovoy <mgurtovoy@nvidia.com>
Subject: [PATCH 2/3] virito_pci: add timeout to reset device operation
Date: Wed, 7 Apr 2021 12:09:23 +0000
Message-ID: <20210407120924.133294-2-mgurtovoy@nvidia.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210407120924.133294-1-mgurtovoy@nvidia.com>
References: <20210407120924.133294-1-mgurtovoy@nvidia.com>
MIME-Version: 1.0
X-EOPAttributedMessage: 0
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: a461cc55-12a7-4fe6-e035-08d8f9bdff4a
X-MS-TrafficTypeDiagnostic: BYAPR12MB4696:
X-Microsoft-Antispam-PRVS: 
 <BYAPR12MB4696AE1413ABD7B9F4D77BEFDE759@BYAPR12MB4696.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:7691;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 0IQ7vUdJzmkK5SCq9PFASTKGxUQRjdF/MLzWvow4nmQ/pt5IOyDCr+J7IUmhZT86MTyJSP+yq2VJh5JXAM+jJquPfBTdRiUCohyclPA9eldWOXxsvU1NkVnffrvLzRLmGhp6P13lrwPKmXCsubzkJyZum7JLWbVCFV2k4/mQllD6mg0A5dpyEbr5h+DL+f18W5PdKbSkX/G7ro5BbuxtatwXsA0gATftJbFcNfhGOAwxUcZ7TGTnIkylZALMFW5agNb+AJ//yBpUQ+MDu9bjmHM4par0YnXLbC9M2tUoJIwd/Fslbj3ZC4FFtDtTdZ5hsngVw6K1T8qW9gri54K6mz1mS7f4LyOggRs9Fl8Dmh8xJZZjp2npeYVGlhoR1Y6UY/Tu0vY0C8u4BQkPCM3ot7L6Z0lH6V7xpovirtbEA/+BZsolAD9/DKw1ZkEEixTMzKuQ5kF3tO/HAcc+6YG+RM3PkRpTXMUux2+x+UVZhDCOVGcUEJaie7l3VCdG+4DgwJ4Urvbg1jZ5SKT3kXjBMsS7uzinNCQm/pR+pcwP/3SNXK3saPfB8FlGdBjPOsgH8cASIoRMo+/o0MLPdTEf3RHkbF2CQZpbVrJyJzNX2JV/2N6IruzAWyG2brVmYtmLf/zGaonE46vYWo3nFqxdHU9vt5SvCKrJGxcHodfoD+8=
X-Forefront-Antispam-Report: 
 CIP:216.228.112.35;CTRY:US;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:mail.nvidia.com;PTR:schybrid02.nvidia.com;CAT:NONE;SFS:(4636009)(346002)(136003)(396003)(376002)(39860400002)(36840700001)(46966006)(70206006)(5660300002)(2906002)(8676002)(8936002)(47076005)(82310400003)(86362001)(336012)(26005)(6666004)(1076003)(186003)(70586007)(4326008)(110136005)(478600001)(54906003)(36906005)(356005)(316002)(83380400001)(426003)(82740400003)(107886003)(2616005)(7636003)(36756003)(36860700001);DIR:OUT;SFP:1101;
X-OriginatorOrg: Nvidia.com
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 07 Apr 2021 12:09:29.9098
 (UTC)
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 a461cc55-12a7-4fe6-e035-08d8f9bdff4a
X-MS-Exchange-CrossTenant-Id: 43083d15-7273-40c1-b7db-39efd9ccc17a
X-MS-Exchange-CrossTenant-OriginalAttributedTenantConnectingIp: 
 TenantId=43083d15-7273-40c1-b7db-39efd9ccc17a;Ip=[216.228.112.35];Helo=[mail.nvidia.com]
X-MS-Exchange-CrossTenant-AuthSource: 
 DM6NAM11FT004.eop-nam11.prod.protection.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Anonymous
X-MS-Exchange-CrossTenant-FromEntityHeader: HybridOnPrem
X-MS-Exchange-Transport-CrossTenantHeadersStamped: BYAPR12MB4696
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

According to the spec after writing 0 to device_status, the driver MUST
wait for a read of device_status to return 0 before reinitializing the
device. In case we have a device that won't return 0, the reset
operation will loop forever and cause the host/vm to stuck. Set timeout
for 3 minutes before giving up on the device.

Signed-off-by: Max Gurtovoy <mgurtovoy@nvidia.com>
---
 drivers/virtio/virtio_pci_modern.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/virtio/virtio_pci_modern.c b/drivers/virtio/virtio_pci_modern.c
index cc3412a96a17..dcee616e8d21 100644
--- a/drivers/virtio/virtio_pci_modern.c
+++ b/drivers/virtio/virtio_pci_modern.c
@@ -162,6 +162,7 @@ static int vp_reset(struct virtio_device *vdev)
 {
 	struct virtio_pci_device *vp_dev = to_vp_device(vdev);
 	struct virtio_pci_modern_device *mdev = &vp_dev->mdev;
+	unsigned long timeout = jiffies + msecs_to_jiffies(180000);
 
 	/* 0 status means a reset. */
 	vp_modern_set_status(mdev, 0);
@@ -169,9 +170,16 @@ static int vp_reset(struct virtio_device *vdev)
 	 * device_status to return 0 before reinitializing the device.
 	 * This will flush out the status write, and flush in device writes,
 	 * including MSI-X interrupts, if any.
+	 * Set a timeout before giving up on the device.
 	 */
-	while (vp_modern_get_status(mdev))
+	while (vp_modern_get_status(mdev)) {
+		if (time_after(jiffies, timeout)) {
+			dev_err(&vdev->dev, "virtio: device not ready. "
+				"Aborting. Try again later\n");
+			return -EAGAIN;
+		}
 		msleep(1);
+	}
 	/* Flush pending VQ/configuration callbacks. */
 	vp_synchronize_vectors(vdev);
 	return 0;

From patchwork Wed Apr  7 12:09:24 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Max Gurtovoy <mgurtovoy@nvidia.com>
X-Patchwork-Id: 12187929
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.8 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 37559C433ED
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:45 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 05F2861153
	for <kvm@archiver.kernel.org>; Wed,  7 Apr 2021 12:09:45 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1352008AbhDGMJw (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Wed, 7 Apr 2021 08:09:52 -0400
Received: from mail-dm6nam10on2061.outbound.protection.outlook.com
 ([40.107.93.61]:33952
        "EHLO NAM10-DM6-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S235268AbhDGMJn (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 7 Apr 2021 08:09:43 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=bxxoBuyTwJOuHATdM+7HqOUlOW1LECbVSQHOWu+xFVQzYTtF89wDcWCRruNhddOIvv5ucpOU3BP8xFMpzt6L2yk434z/YgLeoX/1HsEMoYJhnrryq8gdE2Zll666/wZfh5aXPmjEQ51T3r7/2Wb/mrZKELfr3WxDR4dkMZj6ImCmVm1VaQf4CfSzgDzOcQgnodg6UE8FTlNzHZ12NGnk2fGD5p6MFJyFR4XYaOyUTigPU19M0PAG18CrybCM9gdxK9jlMUk1I08O9zaU4td8SNh6h65Qv5aLgap96NHas2dw4VBR5bQWyMk1I8aINQ/Fi6hZ1RMdzF97ffEwwKBDpw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=amHnrLvTvWjsynprryF/7TQPxzXdsSUM7yMqWphDQTs=;
 b=lEW2EqL8RBPy6gAbIkeA5VerDuuU7xdZDxRB+kp68NleA/YWZjHJzix2yDsK5rZkC30PE/2KVgCuDJI0z73fEZqEWyfzKs+pB0tJNptMcCxLpCxdQpjjRbix1fC2RLjCl2xLR5JgbmpmgJR8j7NHjCspdBTEO7SOWgRWchHf51MqtWIOjnZNCDTRKX6py1FSXT6Ygy/abQfway3/0Woe9d+llF6Dmm2+xKXPcy4OHBR6JIbH5CRz9EmoybUFVmnQqZDMEQIvNjS3ODaavCUOTaOPO1gPZwUftMeO1+4UzgF4BoNMIhYKQ9cUwPDFPa0X1pqBVG2CoQw9SXyFnehk8Q==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass (sender ip is
 216.228.112.36) smtp.rcpttodomain=lists.linux-foundation.org
 smtp.mailfrom=nvidia.com; dmarc=pass (p=none sp=none pct=100) action=none
 header.from=nvidia.com; dkim=none (message not signed); arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=Nvidia.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=amHnrLvTvWjsynprryF/7TQPxzXdsSUM7yMqWphDQTs=;
 b=b5PyGbzNXVRIW/YlViiwT4Zogwnp83mLxZ7g18t66E1fgmPRUgMrGLuU9nkPvqrScugksAIrE5TphFCJhuf4daQbpJPJhDj3frgn0wyvJtm2ALFHw+DnRsoDM7wJj5joHBlGKXDoaV7OF85GpQ39li9APkk+57frIHmoueUUSvpo60r/sLJejOlX+ZpNfF7d3EJZThW40nqahuMbDlxO6sjhIAtunnnmj0iuh+IES6yJ5NzIODjjqOuxHmiqfKv1RbedGgHYamKyF+PnKj+gYdCOKtPQJrkp1TkvQr5tVXe7NuyVeDwFY2jwb7MwnyCPTidZJAfClKaV0G58mek+Wg==
Received: from MW4PR03CA0140.namprd03.prod.outlook.com (2603:10b6:303:8c::25)
 by BN6PR1201MB2480.namprd12.prod.outlook.com (2603:10b6:404:b0::9) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3999.32; Wed, 7 Apr
 2021 12:09:32 +0000
Received: from CO1NAM11FT025.eop-nam11.prod.protection.outlook.com
 (2603:10b6:303:8c:cafe::ee) by MW4PR03CA0140.outlook.office365.com
 (2603:10b6:303:8c::25) with Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4020.16 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:32 +0000
X-MS-Exchange-Authentication-Results: spf=pass (sender IP is 216.228.112.36)
 smtp.mailfrom=nvidia.com; lists.linux-foundation.org; dkim=none (message not
 signed) header.d=none;lists.linux-foundation.org; dmarc=pass action=none
 header.from=nvidia.com;
Received-SPF: Pass (protection.outlook.com: domain of nvidia.com designates
 216.228.112.36 as permitted sender) receiver=protection.outlook.com;
 client-ip=216.228.112.36; helo=mail.nvidia.com;
Received: from mail.nvidia.com (216.228.112.36) by
 CO1NAM11FT025.mail.protection.outlook.com (10.13.175.232) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384) id
 15.20.4020.17 via Frontend Transport; Wed, 7 Apr 2021 12:09:32 +0000
Received: from HQMAIL101.nvidia.com (172.20.187.10) by HQMAIL101.nvidia.com
 (172.20.187.10) with Microsoft SMTP Server (TLS) id 15.0.1497.2; Wed, 7 Apr
 2021 12:09:31 +0000
Received: from r-nvmx02.mtr.labs.mlnx (172.20.145.6) by mail.nvidia.com
 (172.20.187.10) with Microsoft SMTP Server id 15.0.1497.2 via Frontend
 Transport; Wed, 7 Apr 2021 12:09:29 +0000
From: Max Gurtovoy <mgurtovoy@nvidia.com>
To: <mst@redhat.com>, <kvm@vger.kernel.org>,
        <virtualization@lists.linux-foundation.org>, <jasowang@redhat.com>
CC: <oren@nvidia.com>, <nitzanc@nvidia.com>,
        Max Gurtovoy <mgurtovoy@nvidia.com>
Subject: [PATCH 3/3] virtio_pci: add module param for reset_timeout
Date: Wed, 7 Apr 2021 12:09:24 +0000
Message-ID: <20210407120924.133294-3-mgurtovoy@nvidia.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210407120924.133294-1-mgurtovoy@nvidia.com>
References: <20210407120924.133294-1-mgurtovoy@nvidia.com>
MIME-Version: 1.0
X-EOPAttributedMessage: 0
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: bf624763-f36e-4851-7413-08d8f9be0094
X-MS-TrafficTypeDiagnostic: BN6PR1201MB2480:
X-Microsoft-Antispam-PRVS: 
 <BN6PR1201MB2480B95685E51CC990315753DE759@BN6PR1201MB2480.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4303;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 EzURaMStfhttVJIAopiXAvQauRS/pXOYV/hxUufEDb8j/PFDX/oxBUz7PeiiT9KvYJvMxKRW3GSzH3qJYIvI2i+B4P41KHMRK3RVSA7lopV7hig0JMe8X4cVCYhjzpC/qSPUX0s+2HboYDVUaGWmp3UR8SX7OzV24+nyLosAKPSRcl6Mqceu72nSE4eC7w3MGHqL6lReIvHnnUFnjRR7FQeJXvGAd5k9nJXRADIYedRg3hMCpZgHMn3y4ZPCqKh1Wv3UmH7NLjFA44F1eYCgMLQMa9PPPvDLyEAsWCj8Ev8PQZMWP4Jrgr8eFSOd/ANnvfxmTITAPgu5QzcyWeoUWXqpvNwuQnWT0JlnLjxbnRsx3y4Cm/UzsojU76EuhgbsZzVChBuM1ZLmlsuEsnrcdveXdYNWhRNbaDWFo6itDX1FWLoKUWVf/RfWNybbu6Qy7svxPdmi5xNjf8G106hq9Cl3ib5ib5caEAoLH3X8E6WERcNCGCdKOlAcQog2HxPQ5OnKxqH+Au+M9CaeaBWHEh74a6ms80OYPHJ0V4JrIpD0Un5xVFTYjxS7WhSn1XiZ5UNzh0KntUHe0hFUxOPnBB+ofjvKqUwfStq8QHI/ZQjWhL5zGLpDltV7KGTyaIb2sI6zg40S6Of5RTam3JFsG4ynM7sAfgfchuOkEnICbVM=
X-Forefront-Antispam-Report: 
 CIP:216.228.112.36;CTRY:US;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:mail.nvidia.com;PTR:schybrid05.nvidia.com;CAT:NONE;SFS:(4636009)(396003)(346002)(136003)(39860400002)(376002)(36840700001)(46966006)(26005)(356005)(36756003)(2906002)(426003)(7636003)(6666004)(70586007)(110136005)(4326008)(336012)(82310400003)(186003)(70206006)(1076003)(86362001)(54906003)(36860700001)(316002)(36906005)(82740400003)(47076005)(83380400001)(478600001)(2616005)(5660300002)(8936002)(8676002)(107886003);DIR:OUT;SFP:1101;
X-OriginatorOrg: Nvidia.com
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 07 Apr 2021 12:09:32.0777
 (UTC)
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 bf624763-f36e-4851-7413-08d8f9be0094
X-MS-Exchange-CrossTenant-Id: 43083d15-7273-40c1-b7db-39efd9ccc17a
X-MS-Exchange-CrossTenant-OriginalAttributedTenantConnectingIp: 
 TenantId=43083d15-7273-40c1-b7db-39efd9ccc17a;Ip=[216.228.112.36];Helo=[mail.nvidia.com]
X-MS-Exchange-CrossTenant-AuthSource: 
 CO1NAM11FT025.eop-nam11.prod.protection.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Anonymous
X-MS-Exchange-CrossTenant-FromEntityHeader: HybridOnPrem
X-MS-Exchange-Transport-CrossTenantHeadersStamped: BN6PR1201MB2480
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Enable the user to set the time for waiting for successful reset by the
virtio controller. Set the default to 180 seconds.

Signed-off-by: Max Gurtovoy <mgurtovoy@nvidia.com>
---
 drivers/virtio/virtio_pci_common.c | 5 +++++
 drivers/virtio/virtio_pci_common.h | 2 ++
 drivers/virtio/virtio_pci_modern.c | 3 ++-
 3 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/virtio/virtio_pci_common.c b/drivers/virtio/virtio_pci_common.c
index 222d630c41fc..3a4c57839ed8 100644
--- a/drivers/virtio/virtio_pci_common.c
+++ b/drivers/virtio/virtio_pci_common.c
@@ -24,6 +24,11 @@ MODULE_PARM_DESC(force_legacy,
 		 "Force legacy mode for transitional virtio 1 devices");
 #endif
 
+unsigned int reset_timeout = 180;
+module_param_named(reset_timeout, reset_timeout, uint, 0644);
+MODULE_PARM_DESC(reset_timeout,
+		 "timeout in seconds for reset virtio device operation");
+
 /* wait for pending irq handlers */
 void vp_synchronize_vectors(struct virtio_device *vdev)
 {
diff --git a/drivers/virtio/virtio_pci_common.h b/drivers/virtio/virtio_pci_common.h
index beec047a8f8d..4760cdf74961 100644
--- a/drivers/virtio/virtio_pci_common.h
+++ b/drivers/virtio/virtio_pci_common.h
@@ -29,6 +29,8 @@
 #include <linux/highmem.h>
 #include <linux/spinlock.h>
 
+extern unsigned int reset_timeout;
+
 struct virtio_pci_vq_info {
 	/* the actual virtqueue */
 	struct virtqueue *vq;
diff --git a/drivers/virtio/virtio_pci_modern.c b/drivers/virtio/virtio_pci_modern.c
index dcee616e8d21..811fc1719d8c 100644
--- a/drivers/virtio/virtio_pci_modern.c
+++ b/drivers/virtio/virtio_pci_modern.c
@@ -162,7 +162,8 @@ static int vp_reset(struct virtio_device *vdev)
 {
 	struct virtio_pci_device *vp_dev = to_vp_device(vdev);
 	struct virtio_pci_modern_device *mdev = &vp_dev->mdev;
-	unsigned long timeout = jiffies + msecs_to_jiffies(180000);
+	unsigned long timeout = jiffies +
+		msecs_to_jiffies(reset_timeout * 1000);
 
 	/* 0 status means a reset. */
 	vp_modern_set_status(mdev, 0);
